source/transfer_list.rst

```diff
@@ -902,5 +902,42 @@ This manifest contains the SPMC attribute node consumed by the SPMD
      - hdr_size
      - Holds a SPMC manifest image in DT format.
 
+
+**Executable entry point information entry layout (XFERLIST_EXEC_EP_INFO)**
+
+This entry type holds an array of `entry_point_info_t` structures.
+`entry_point_info_t` is a data structure containing information used by a
+bootloader to initialize the execution state of an image. This entry provides
+the consumer with all information to execute subsequent images.
+
+.. _tab_entry_point_info:
+.. list-table:: Entry point info type layout
+   :widths: 2 2 2 8
+
+   * - Field
+     - Size (bytes)
+     - Offset (bytes)
+     - Description
+
+   * - tag_id
+     - 0x3
+     - 0x0
+     - The tag_id field must be set to **0x102**.
+
+   * - hdr_size
+     - 0x1
+     - 0x3
+     - |hdr_size_desc|
+
+   * - data_size
+     - 0x4
+     - 0x4
+     - The size of SPMC manifest in bytes.
+
+   * - ep_info
+     - data_size
+     - hdr_size
+     - Holds an array of `entry_point_info_t` structures.
```

Comment:
This struct should be defined here. The spec should stand along, so we need to know what is in it.

Also, why do we want an array of these? I suspect that is fine, but I just don't know why...

Reply:
There are boot-loaders responsible for the execution of more than one image. For example, BL31 in some configurations will execute BL32, RMM and a non-secure image. Packing the information in an array, allows several entry points to be shared with the next stage in a single TE. We would otherwise require several TE's to convey the same information.

Reply:
Probably using several TEs is better

Reply:
@raymo200915 raised this in the corresponding TF-A patch. Does the specification allow for several entries of the same TE type in a TL?

Reply:
The spec doesn't tell explicitly whether multiple TEs can share same tag ID or I am wrong, but tag ID is the only field that can identify a TE at the moment.
Shall we consider to add another field in the TE header, so that we can identify a TE even when its tag ID is not unique in a TL?

Reply:
> The spec doesn't tell explicitly whether multiple TEs can share same tag ID or I am wrong, but tag ID is the only field that can identify a TE at the moment. 
Even though spec does not explicitly mentions it, IMHO we should avoid using same tag ID for multiple TEs. This will keep things simple (e.g the searching of a tag ID returns the first occurrence) 

> Shall we consider to add another field in the TE header, so that we can identify a TE even when its tag ID is not unique in a TL?
IMO we should avoid adding a new field just to cater for a specific case (until its unavoidable).  As we know that there are going to be finite number of TEs (less than 10) and we have enough range to accommodate it in tag ID.
In this particular case the maximum array size would be 3 (BL32/BL33, RMM). So even if we decide to have several TEs we can easily provide 3 tag IDs in Arm range.

Reply:
> The spec doesn't tell explicitly whether multiple TEs can share same tag ID

It does (now) actually, see the "TL Contents" section. I don't think it forbids putting an array in a single tag, though. In my opinion, this probably depends on whether these are generated all in one go or a different points in time. If you know for sure that the passing stage will always add all of these entry points at once, I think an array would be okay. If you think there's a chance that one stage might add some entry points and then later either the same stage or a later one might want to add some more, then using a separate TE for each would be a good idea.

I'm not really sure what you mean by "identify a TE"? Within the context of a single firmware stage, a TE can be identified by its address, of course (e.g. to pass around between APIs). Across firmware stages I don't see why we would want to identify TEs, since every TE should be independent and not reference other TEs. When parsing a TL to find something (e.g. your entry points here), I would suggest an API that searches for all TEs of a certain tag ID and then returns each of them in turn (e.g. with a callback or a find_next() function) so the caller can figure out which one(s) it is interested in.

Reply:
>  this probably depends on whether these are generated all in one go or a different points in time

In context of current PR, these entries are all generated at once in BL2 stage.

Reply:
I would prefer to have different TE with different ID for each ep_info instead of putting all ep_info into an array, just in order to simply the scenario in BL31 - if we still have sufficient spare IDs.

> I'm not really sure what you mean by "identify a TE"? 
I mean the operations to locate a specific TE. For example, search a TE by a tag ID or anything else and returns an TE address. Anyway, if we are not going to have multiple TEs with same tag ID, we don't have the worries.

Reply:
Chapter 'TL Contents'  has below statement -

> Tags are expected to have a simple layout (representable by a C structure) and each tag should only 
> represent data for a single logical concept. Data for multiple distinct concepts should be split across
> different tags

IIUC, it signifies that a single or multiple sets of data can be denoted by the same tag-ID, as long as they encapsulate the same logical entity. In this specific scenario, the entry point is a singular logical entity, thus portraying it in either a singular (single entry-point info) or multiple fashion (multiple entry-point infos) should be considered acceptable.
If above understanding is correct then presenting same logical information through multiple tags could potentially create confusion.

Reply:
If re-usability is a stated design goal of the specification, our options are either to maintain the TE as described in this patch, or to more narrowly define it so that each instance of the structure gets its own entry. From @ManishVB-Arm message it seems either is viable, although, some rework might be required in the implementation used by TF-A to support having multiple instances of the same TE. 
>The same tag ID may occur multiple times in the TL to represent multiple instances of the same kind of object.

On the other suggestion, I think it would be misguided to create tags for each type of `entry_point_info_t`. Having multiple representations of the same data risks creating confusion around usage. This could result in a scenario where every platform/chipset has to re-invent the wheel for their uses-case because of a lack of a generic representation.

Reply:
@sjg20 4-bytes of `param_header_t` are reserved to provide the attributes of the represented structure. In fact, we already use bits 0 and 5 of that field to encode the security state. At the moment, I'm using those bits as a proxy for the image id. That is enough for the reference platform I'm working on, although, there's no reason we can't add an additional image id field in the attributes.